### PR TITLE
Windows MR support, WIP

### PR DESCRIPTION
**Pull request by caseyyee**

### CODE REVIEW

**Comment by cvan** [comment]
nice work!

what's left (WIP)?

**Comment by caseyyee** [comment]
Just fixing issues with not getting the headset projection and view matrices for some reason.

File: ProjectSettings/ProjectSettings.asset [hunk 1]
```diff
@@ -66,7 +66,7 @@ PlayerSettings:
   defaultIsFullScreen: 1
   defaultIsNativeResolution: 1
   macRetinaSupport: 1
-  runInBackground: 0
+  runInBackground: 1
```
[the remark this answers is not in the document]
Comment (caseyyee) [reply ProjectSettings/ProjectSettings.asset]: With Edge, when you enter WebVR mode, the window focus is taken away.   Unity player pauses the game when window is blurred, this works around that.   Might be a way to pass it into the gameInstance in index.html?

**Comment by caseyyee** [comment]
rebased and working now.
@cvan r?

**Comment by cvan** [comment]
I tested this was from several hours ago, and it didn't work on my Dell Visor.

Here are some screenshots of issues I had (look towards the higher-numbered ones): https://drive.google.com/open?id=1ylF13GeSoi4PgFjJ7F7tuLSIeUBTZy5s

I'm going to pull again and test again. Thanks!

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 2]
```diff
@@ -242,23 +211,50 @@
...
-  function init() {
-    // shim raf so that we can drive framerate using VR display.
-    window.requestAnimationFrame = onRequestAnimationFrame;
-    // messages from Unity.
-    document.addEventListener('Unity', onUnity);
-    window.addEventListener('resize', onResize, true);
-    document.addEventListener('keydown', onKeyDown);
-    entervrButton.addEventListener('click', onEnterVR);
+  if (navigator.getVRDisplays) {
+    frameData = new VRFrameData();
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/webvr.js]: can you check for `'VRFrameData' in window`?

File: Build/webvr.js [hunk 3]
```diff
@@ -242,23 +211,50 @@
...
-    entervrButton.addEventListener('click', onEnterVR);
+  if (navigator.getVRDisplays) {
+    frameData = new VRFrameData();
+
+    navigator.getVRDisplays().then(function(displays) {
+      if (displays.length > 0) {
+        vrDisplay = displays[displays.length - 1];
+
+        // check to see if we are polyfilled
+        if (vrDisplay.displayName.indexOf('polyfill') > 0) {
```
Comment (cvan) [inline Build/webvr.js]: can you change this to [match this check](https://github.com/caseyyee/unity-webvr-export/pull/53/files#diff-81e5be25454a5c03ab739760c4b399b5L33)? (feel free to create an `isVRDisplayPolyfilled` function)

File: Build/webvr.js [hunk 4]
```diff
@@ -242,23 +211,50 @@
...
+          showInstruction(document.querySelector('#novr'));
+        } else {
+          status.dataset.enabled = true;
```
Comment (cvan) [inline Build/webvr.js]: should this be the string `'true'` like [the one below](https://github.com/caseyyee/unity-webvr-export/pull/53/files#diff-81e5be25454a5c03ab739760c4b399b5R241)?

File: ProjectSettings/ProjectSettings.asset (hunk 1 above)
[the remark this answers is not in the document]
Comment (cvan) [reply ProjectSettings/ProjectSettings.asset]: ah, gotcha - thanks

File: Build/webvr.js [hunk 5]
```diff
@@ -68,131 +35,133 @@
...
-      rightViewMatrix: Array.from(rightViewMatrix),
-      sitStand : Array.from(sitStand),
-      controllers: vrGamepads
-    };
+        var vrData = {
+          leftProjectionMatrix: Array.from(leftProjectionMatrix),
+          rightProjectionMatrix: Array.from(rightProjectionMatrix),
+          leftViewMatrix: Array.from(leftViewMatrix),
+          rightViewMatrix: Array.from(rightViewMatrix),
+          sitStand : Array.from(sitStand),
```
Comment (cvan) [inline Build/webvr.js]: remove space before `:`

File: Build/webvr.js [hunk 6]
```diff
@@ -68,131 +35,133 @@
...
+        }
+        mat4.transpose(sitStand, sitStand);
+
+        // gamepads
+        gamepads = navigator.getGamepads();
+        vrGamepads = [];
+        for (var i = 0; i < gamepads.length; ++i) {
+          var gamepad = gamepads[i];
+          if (gamepad) {
+            if (gamepad.pose || gamepad.displayId) {
```
Comment (cvan) [inline Build/webvr.js]: can you combine this with the statement above?

File: Build/webvr.js [hunk 7]
```diff
@@ -68,131 +35,133 @@
...
+        if (vrDisplay.stageParameters) {
+          mat4.copy(sitStand, vrDisplay.stageParameters.sittingToStandingTransform);
+        } else {
+          mat4.identity(sitStand);
+          mat4.translate(sitStand, sitStand, [0, defaultHeight, 0]);
+        }
+        mat4.transpose(sitStand, sitStand);
+
+        // gamepads
+        gamepads = navigator.getGamepads();
```
Comment (cvan) [inline Build/webvr.js]: can you check for the presence of `'getGamepads' in navigator`?

**Comment by cvan** [comment]
@caseyyee: Can you try upgrading your version of Unity (from `2017.2.0f3` to `2017.3.0f3`)? I'm not confident that's the issue, but it's worth trying.

Are you testing this with Vive, Rift, and/or Samsung Odyssey (Windows MR)?

Thanks!

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 2 above)
Comment (caseyyee) [reply Assets/WebGLTemplates/WebVR/webvr.js]: Can we not assume that `VRFrameData` exists if `getVRDisplays`?

https://github.com/toji/webvr.info/blob/master/samples/05-room-scale.html#L194

Doesn't seem to do it either.

File: Build/webvr.js (hunk 3 above)
Comment (caseyyee) [reply Build/webvr.js]: Ah yes, i wiped that that change out by accident when rebasing.  thanks

**Comment by caseyyee** [comment]
@cvan you want to give this another go?

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 8]
```diff
@@ -220,30 +219,35 @@
...
-  } else {
-    console.log('Your browser does not support WebVR!');
+      navigator.getVRDisplays().then(function(displays) {
+        if (displays.length > 0) {
+          vrDisplay = displays[displays.length - 1];
+
+          // Check to see if we are polyfilled.
+          var isPolyfilled = (vrDisplay.deviceId || '').indexOf('polyfill') > 0 ||
+            (vrDisplay.displayName || '').indexOf('polyfill') > 0 ||
+            (vrDisplay.deviceName || '').indexOf('polyfill') > 0 ||
```
Comment (caseyyee) [inline Assets/WebGLTemplates/WebVR/webvr.js]: @cvan I had to add in `deviceName`, is `displayName` used anywhere at all?  or did you mean to actually write in `deviceName`?

**Comment by caseyyee** [comment]
Going to merge for now since it seems to be working for me.  if @cvan you could give another testing pass, that would be great.

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 8 above)
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/webvr.js]: https://github.com/immersive-web/cardboard-vr-display/blob/ee3b5aa/dist/cardboard-vr-display.js#L2023

you can ignore [`displayName`](https://github.com/immersive-web/cardboard-vr-display/blob/ee3b5aa/dist/cardboard-vr-display.js#L2827); that's deprecated

**Comment by cvan** [comment]
## Edge
- works fantastically! controllers work (a bit finicky but the best of the three browsers), and I'm positioned properly.
- this is the best performance-wise! 👍 

## Chrome Canary
- flags needed: `chrome://flags#enable-webvr`, `chrome://flags#openvr`, `chrome://flags#enable-gamepad-extensions`
- after "Loading..." screen, the **scene is black**. (until I press the "Enter VR" button, there are constant **Console errors**: `Screen position out of view frustum    blob:https://caseyye…c-66cfdf698c02:7636 `
- when I click the "Enter VR" button, the left/right mirrored canvas appears as expected.
- the controllers are visually very finicky (flickering and their positions moving around quite a bit). if I keep my hand still, the hand is constantly shifting all over.
- otherwise, performs quite well! and I'm positioned properly.

## Firefox
- runs now! (I was originally spawned at the center of the floor, but when I invoked the Steam Windows MR option for "Reset Seating Position," my position reverted to be okay.)
- controllers are a bit finicky, but not as finicky as they are in Chrome Canary. but still very bad. if I keep my hand still, the hand is constantly shifting all over.

**Comment by caseyyee** [comment]
@cvan thanks for the test.   I didn't check Canary.  I'll give that a go to see if I can troubleshoot the issue.

**Pull request merged**